services/gundeck/src/Gundeck/API.hs

```diff
@@ -131,7 +137,94 @@ sitemap = do
 
 type JSON = Media "application" "json"
 
-docs :: ByteString ::: JSON -> Gundeck Response
-docs (url ::: _) =
+docsH :: ByteString ::: JSON -> Gundeck Response
+docsH (url ::: _) =
   let doc = mkSwaggerApi (decodeLatin1 url) Model.gundeckModels sitemap
    in return $ json doc
+
+addTokenH :: UserId ::: ConnId ::: JsonRequest PushToken ::: JSON -> Gundeck Response
+addTokenH (uid ::: cid ::: req ::: _) = do
+  newtok <- fromJsonBody req
+  handleAddTokenResponse <$> Push.addToken uid cid newtok
+
+handleAddTokenResponse :: Push.AddTokenResponse -> Response
+handleAddTokenResponse = \case
+  Push.AddTokenSuccess newtok -> success newtok
+  Push.AddTokenNoBudget -> snsThreadBudgetReached
+  Push.AddTokenNotFound -> notFound
+  Push.AddTokenInvalid -> invalidToken
+  Push.AddTokenTooLong -> tokenTooLong
+  Push.AddTokenMetadataTooLong -> metadataTooLong
+
+success :: PushToken -> Response
+success t =
+  let loc = Text.encodeUtf8 . tokenText $ t ^. token
+   in json t & setStatus status201 & addHeader hLocation loc
+
+invalidToken :: Response
+invalidToken =
+  json (Error status400 "invalid-token" "Invalid push token")
+    & setStatus status404
+
+snsThreadBudgetReached :: Response
+snsThreadBudgetReached =
+  json (Error status400 "sns-thread-budget-reached" "Too many concurrent calls to SNS; is SNS down?")
+    & setStatus status413
+
+tokenTooLong :: Response
+tokenTooLong =
+  json (Error status400 "token-too-long" "Push token length must be < 8192 for GCM or 400 for APNS")
+    & setStatus status413
+
+metadataTooLong :: Response
+metadataTooLong =
+  json (Error status400 "metadata-too-long" "Tried to add token to endpoint resulting in metadata length > 2048")
+    & setStatus status413
+
+notFound :: Response
+notFound = empty & setStatus status404
+
+deleteTokenH :: UserId ::: Token ::: JSON -> Gundeck Response
+deleteTokenH (uid ::: tok ::: _) = setStatus status204 empty <$ Push.deleteToken uid tok
+
+listTokensH :: UserId ::: JSON -> Gundeck Response
+listTokensH (uid ::: _) = setStatus status200 . json <$> Push.listTokens uid
+
+pushH :: Request ::: JSON -> Gundeck Response
+pushH (req ::: _) = do
+  ps <- fromJsonBody (JsonRequest req)
+  empty <$ Push.push ps
+
+paginateH :: JSON ::: UserId ::: Maybe ByteString ::: Maybe ClientId ::: Range 100 10000 Int32 -> Gundeck Response
+paginateH (_ ::: uid ::: sinceRaw ::: clt ::: size) = do
+  Notification.PaginateResult gap page <- Notification.paginate uid (join since) clt size
+  pure . updStatus gap . json $ page
+  where
+    since :: Maybe (Maybe NotificationId)
+    since = parseUUID <$> sinceRaw
```

Comment:
how about :

```
since :: Maybe NotificationId
since = parseUUID =<< sinceRaw

```

Reply:
and think we can just inline  it in the call to `Notification.paginate` then.

join and fmap is the same as bind

Reply:
nope, won't work, because of the bizarre semantics of this end-point.  there is two uses of `since`; one joins, and the other checks for a parse error.  they do different things.

Reply:
I disagree with your reading of this code.    Both lead to a 404. There's just some confusing code duplication that hides this fact